### ISSUE TITLE
Implement "intersection lifetimes" to improve diagnostics and accept rare impl trait cases

### ISSUE THREAD

**Issue by oli-obk** [rust-lang/rust#89394]
TLDR: The following does not compile but it should

```rust
trait Trait<'a, 'b> {}
impl<T> Trait<'_, '_> for T {}

struct Struct<'a>(&'a u8);

fn foo<'a, 'b>(a: Struct<'a>, b: Struct<'b>) -> impl Trait<'a, 'b> {
    if false { a } else { b }
}
```

The lifetime of the `if false { a } else { b }` is `'a` OR `'b`, but right now we can only express `'a` AND `'b` (as `'lifetime: 'a + 'b`).
Beyond diagnostics, only impl trait can actually encounter such lifetimes in a somewhat reasonable way.

So the plan is as follows:

1. add intersection lifetimes to mir borrowck, without actually creating them
2. specifically create them for member constraints that would otherwise be unresolvable
3. put a lot of effort into making the diagnostics comprehensible

Some details: https://rust-lang.zulipchat.com/#narrow/stream/187312-wg-async-foundations/topic/member.20constraints.20PR.20.2389056/near/255496552

Related (purely diagnostics) PR: https://github.com/rust-lang/rust/pull/89327

**Comment by fmease** [comment]
Oh, shouldn't this have been solved with member constraints #84701?

**Comment by theemathas** [comment]
Relevant issues: https://github.com/rust-lang/rust/issues/148401, https://github.com/rust-lang/rust/issues/128752, https://github.com/rust-lang/rust/issues/152775
